parlai/agents/starspace_torch/starspace_torch.py

```diff
@@ -0,0 +1,367 @@
+            '-ecands', '--eval-candidates', type=str, default='inline',
+            choices=['batch', 'inline', 'fixed', 'vocab', 'custom'],
+            help='The source of candidates during evaluation (defaults to the same'
+                 'value as --candidates if no flag is given)')
+        StarspaceTorchAgent.dictionary_class().add_cmdline_args(argparser)
+
+    def __init__(self, opt, shared=None):
+        """Set up model if shared params not set, otherwise no work to do."""
+        super().__init__(opt, shared)
+        opt = self.opt
+        self.reset_metrics()
+        self.id = 'Starspace'
+
+        self.ys_cache = []
+        self.ys_cache_sz = opt['cache_size']
+        self.truncate = opt['truncate'] if opt['truncate'] > 0 else None
+
+        # set up criterion
+        self.criterion = torch.nn.CosineEmbeddingLoss(
+            margin=opt['margin'], size_average=False
+        )
+        self.reset()
+
+    def build_model(self):
+        print("[ creating StarspaceTorchAgent ]")
+        # this is not a shared instance of this class, so do full init
+        if (self.opt.get('model_file') and
+                (os.path.isfile(self.opt.get('model_file') + '.dict')
+                 or (self.opt['dict_file'] is None))):
+            # set default dict-file if not set
+            self.opt['dict_file'] = self.opt['model_file'] + '.dict'
+        # load dictionary and basic tokens & vectors
+        self.dict = DictionaryAgent(self.opt)
+
+        self.model = Starspace(self.opt, len(self.dict), self.dict)
+        if self.opt.get('model_file') and os.path.isfile(self.opt['model_file']):
+            self.load(self.opt['model_file'])
+            self.reset()
+        else:
+            self._init_embeddings()
+        self.model.share_memory()
+        if self.use_cuda:
+            self.model.cuda()
+
+    def train_step(self, batch):
+        """Train on a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        batchsize = batch.text_vec.size(0)
+        self.model.train()
+        self.optimizer.zero_grad()
+
+        if not batch.candidates:
+            print('No negatives yet!')
+            return Output()
+
+        cands, cand_vecs, label_inds = self._build_candidates(
+            batch, source=self.opt['candidates'], mode='train')
+
+        if self.opt.get('input_dropout', 0) > 0:
+            batch_text, batch_label, batch_cand_vecs = self.input_dropout(
+                batch.text_vec,
+                batch.label_vec,
+                cand_vecs
+            )
+            xe, ye = self.model(
+                batch_text,
+                ys=batch_label,
+                cands=batch_cand_vecs
+            )
+        else:
+            xe, ye = self.model(
+                batch.text_vec,
+                ys=batch.label_vec,
+                cands=cand_vecs
+            )
+        xe = torch.cat([xe for _ in range(ye.size(1))], dim=1)
+
+        if self.use_cuda:
```

Comment:
Define y and y_0, then move them to cuda separately. Slightly DRYER. You could also do `.to(xe.device)` like suggested above.